cpp/src/Slice/MetadataValidator.cpp

```diff
@@ -0,0 +1,299 @@
+        }
+
+        if (!metadata->arguments().empty())
+        {
+            return "the 'UserException' metadata does not take any arguments";
+        }
+
+        return nullopt;
+    }
+}
+
+bool
+Slice::MetadataValidator::visitUnitStart(const UnitPtr& p)
+{
+    DefinitionContextPtr dc = p->findDefinitionContext(p->topLevelFile());
+    assert(dc);
+    dc->setMetadata(validate(dc->getMetadata(), p));
+    return true;
+}
+
+bool
+Slice::MetadataValidator::visitModuleStart(const ModulePtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    return true;
+}
+
+void
+Slice::MetadataValidator::visitClassDecl(const ClassDeclPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+}
+
+bool
+Slice::MetadataValidator::visitClassDefStart(const ClassDefPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    return true;
+}
+
+void
+Slice::MetadataValidator::visitInterfaceDecl(const InterfaceDeclPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+}
+
+bool
+Slice::MetadataValidator::visitInterfaceDefStart(const InterfaceDefPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    return true;
+}
+
+bool
+Slice::MetadataValidator::visitExceptionStart(const ExceptionPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    return true;
+}
+
+bool
+Slice::MetadataValidator::visitStructStart(const StructPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    return true;
+}
+
+void
+Slice::MetadataValidator::visitOperation(const OperationPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+}
+
+void
+Slice::MetadataValidator::visitParamDecl(const ParamDeclPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+}
+
+void
+Slice::MetadataValidator::visitDataMember(const DataMemberPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+}
+
+void
+Slice::MetadataValidator::visitSequence(const SequencePtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    p->setTypeMetadata(validate(p->typeMetadata(), p->type()));
+}
+
+void
+Slice::MetadataValidator::visitDictionary(const DictionaryPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    p->setKeyMetadata(validate(p->keyMetadata(), p->keyType()));
+    p->setValueMetadata(validate(p->valueMetadata(), p->valueType()));
+}
+
+void
+Slice::MetadataValidator::visitEnum(const EnumPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+}
+
+void
+Slice::MetadataValidator::visitConst(const ConstPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    p->setTypeMetadata(validate(p->typeMetadata(), p->type()));
+}
+
+MetadataList
+Slice::MetadataValidator::validate(MetadataList metadata, const SyntaxTreeBasePtr& p)
+{
+    // Iterate through the provided metadata and check each one for validity.
+    for (MetadataList::const_iterator i = metadata.begin(); i != metadata.end(); ++i)
+    {
+        const MetadataPtr& meta = *i++;
```

Comment:
Looks like you're incrementing the iterator twice.

Reply:
Good catch. Fixed.
My fingers are too used to typing the `++i` in the for loop...